## Exercise 1

0. First lets load the data

In [2]:
import numpy as np

viajes = np.array([[32, 41, 28],
                  [37, 46, 35],
                  [29, 44, 51],                    
                  [42, 39, 33],                    
                  [48, 52, 36]])

1.
 I don't know why but I keep having this bug in which I want to execute a jupyter notebook inside VSCode using uv in my NixOS WSL but it keeps telling me to install the dependencies even though I already have them in my pyproject.toml​.

What the AI answered to me that the fix is 

```
The fix
[tool.uv] python-preference = "only-managed" in pyproject.toml. uv's interpreters are python-build-standalone builds whose interpreter is /lib64/ld-linux-x86-64.so.2 — so the whole process runs under nix-ld and the wheels resolve. only-managed means it can't silently regress after a rm -rf .venv.
```

What does "nix-ld" mean? It seems it is very important, particularly for the setting of NixOS in WSL... could it be "list directory"? A particular, visible, subset of NixOS that runs a certain environment? 

After the fix that Claude did I can finally work with jupyter notebooks. 

In [3]:
viajes

array([[32, 41, 28],
       [37, 46, 35],
       [29, 44, 51],
       [42, 39, 33],
       [48, 52, 36]])

2. I remember that in R we could label the axis to our liking. As each row represents a day and each column represents a different route, I wonder how could I make so I have a clear attribute in `viajes` that reminds me of this in case I forget it.

3. The current version of this notebook's numpy is greater or equal than 2.5.2. 

In [4]:
viajes.ndim

2

4. I don't know what `ndim` does but it seems it provides me with the number of dimensions. 

In [5]:
viajes.shape

(5, 3)

5. `.shape` seems to give me the number of rows and then columns.

In [6]:
viajes.dtype

dtype('int64')

6. This is the only one which I knew what meant. I don't know what the "d" stands for but I have already used a variety of numpy types such as `float64` and `uint8`. 

I don't understand what "relacione cada eje con el contexto" means, but if it is in the term of coordinates each row describes how much time a route took for each of the days. 

## Exercise 2

In [7]:
indices_dias = np.arange(4)
indices_dias.shape
indices_dias.dtype = np.uint16
indices_dias

/tmp/ipykernel_1410/3646639976.py:3: DeprecationWarning: Setting the dtype on a NumPy array has been deprecated in NumPy 2.5.
Instead of changing the dtype on an array x, create a new array with x.view(new_dtype)
  indices_dias.dtype = np.uint16


array([0, 0, 0, 0, 1, 0, 0, 0, 2, 0, 0, 0, 3, 0, 0, 0], dtype=uint16)

```bash
DeprecationWarning: Setting the dtype on a NumPy array has been deprecated in NumPy 2.5.
Instead of changing the dtype on an array x, create a new array with x.view(new_dtype)
  indices_dias.dtype = np.uint16
```

8. It seems that I cannot change the type of the numpy array. It makes sense to me as that would force coertion, but even then I can't quite understand why it has been deprecated. That coertion seems to be enforced for efficiency, but then I am left with a numpy array unless I make a copy of it? 

`array([0, 0, 0, 0, 1, 0, 0, 0, 2, 0, 0, 0, 3, 0, 0, 0], dtype=uint16)`

I can't quite understand why the array is that long... Also I now see that the upper limit is not inclusive

In [8]:
indices_dias = np.arange(5)
indices_dias

array([0, 1, 2, 3, 4])

10. Now it has seem to work, I don't know why the previous `indices_dias` was that long. I assume in this case where there is a rewritting of the variable the previous one is simply deleted and the new instantiation takes its place. 

In [9]:
reference = np.zeros((5, 3))
reference

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

12. I got `TypeError: Cannot interpret '3' as a data type` so I assume I can instantiate the shape directly with parenthesis

In [10]:
reference = np.zeros((5, 3))
reference

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

13. Indeed my suspicion was correct, now I just need to set the others for 30. I will use ones and a multiplication. 

In [11]:
form = np.ones((5, 3))
form * 30
form

array([[1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.]])

14. I remember that I read that when you did an operation to a numpy array the changes happened to the array itself instead of a copy; that is, they're pased by reference instead of value, however in this case it doesn't work that way. That phenomenon might be only present with slicing

In [12]:
form[:,:] = 30
form

array([[30., 30., 30.],
       [30., 30., 30.],
       [30., 30., 30.],
       [30., 30., 30.],
       [30., 30., 30.]])

15. Indeed it seems that the operations do not change the underlying array, but when it involves slicing it does. 

To be honest I don't see much value in having these arrays filled with zeros and 30s... I could say that they allow scalar multiplication but numpy already does it for you underneath. They might help with matrix multiplication though. 

In [13]:
float_data = np.array([3.7, 1.2])
float_data.astype(np.int32)
float_data

array([3.7, 1.2])

16. Again, there seems to not be a coertion `array([3.7, 1.2])`. 
I will reasign by default for now until I understand when there is coertion and when there is not. 

In [14]:
float_data = float_data.astype(np.int32)
float_data

array([3, 1], dtype=int32)

17. The reason why this operation might be risky is that it always rounds down, so while in some cases the decimals might be too small that they can be truncated without a real loss in meaning, it might as well be near the next whole integer but we treat it as it didn't really matter. Depending on the scale of the data that decimal part might contain more information that we want to get rid of. It is important to note that this truncations compound very heavily in multiplications, getting worse the more multiplications we do. 

## Exercise 3

18. `viajes[2,1]` is the second row and first column...

In [15]:
viajes

array([[32, 41, 28],
       [37, 46, 35],
       [29, 44, 51],
       [42, 39, 33],
       [48, 52, 36]])

... so the values is the lenght of time that the first route took on Tuesday. It has the value 37. 

In [16]:
viajes[2,1]

np.int64(44)

22. I forgot 0-indexing! I was actually one diagonal to the left to the real value. 

`viajes[:,0]` This means grab all the rows on the first column. This is all the times that the first route took in the given week. The expected output should be `[32, 37, 29, 42, 48]`

In [17]:
viajes[:,0]

array([32, 37, 29, 42, 48])

23. I was right this time! The reason why is not a matrix is that while we are grabing all the rows by using `:` we are limiting ourselves to a single column by putting an index. A single column with 5 rows is indistinguishable from a 1D vector with 5 values, that is why we got a horizontal layout instead of a vertical one; they confer the same meaning but another makes a more efficient use of space in the screen. 

24. As we need the two first rows we should make use of `[:1]` assuming the upper bound is inclusive. For B and C we should make use of `[1:]`. The area selected is a 2x2 square that touches the upper right corner of the matrix. 

In [18]:
viajes[:1, 1:]

array([[41, 28]])

25. I got it wrong... Lets check the data again. 

In [19]:
viajes

array([[32, 41, 28],
       [37, 46, 35],
       [29, 44, 51],
       [42, 39, 33],
       [48, 52, 36]])

26. It seems it is not inclusive, but why it isn't selecting all rows? Because that is what I want, I just need to move the expresion of the left

In [20]:
viajes[:2, 1:]

array([[41, 28],
       [46, 35]])

27. Indeed it work. It has a shape of (2, 2)

In [21]:
viajes.shape

(5, 3)

27. Now even with slicing it didn't change... It makes sense. That I want to isolate part of the data doesn't mean that is the final version of how I want it to look. It seems that the change with slicing is only when I am assigning new data. 

In [22]:
sub_array = viajes[:2, 1:]
sub_array.shape

(2, 2)

28. I don't know what it refers to "indexacion avanzada" but I imagine I can concatenate several numpy arrays in order to get the data in the order asked. It is not a continous range for 3 reasons: 
1. The data in here is intrinsecally discrete (both by the data type of integers and also the mere fact that it jumps from one day to the other; the samples are not being taken constantly)
2. We're breaking the chronological order as we're putting Friday first, Monday next and Wendesday last. 
3. Even if Monday, Wendesday and Friday were in the correct order we still have jumps in the data as we lack Tuesday and Thursday (we might even go that far to say we lack Saturday and Sunday too).

In [25]:
friday = viajes[4, :]
monday = viajes[0, :]
wednesday = viajes[2, :]
new_order = friday + monday + wednesday
new_order

array([109, 137, 115])

31. That was not the output that I expected! `array([109, 137, 115])`. What I want is that the 3 separate rows join in that specific order. Now that I have checked it seems that advanced indexing is a way to obtain particular matrices by making use of integers or a mask. Maybe I can get the matrix requested by simply putting the order in which I want it

In [27]:
new_order = viajes[[4, 0, 2]]
new_order

array([[48, 52, 36],
       [32, 41, 28],
       [29, 44, 51]])

32. It worked! it is important to note that there must be another square brackets inside the ones that accompany `viajes`, as without them it told me that I was trying to access 3 dimensions where my numpy array just had 2. 

## Exercise 4

In [ ]:
bloque = viajes[:2, :2]
previous_value = viajes[0, 0]
bloque[0, 0] = 999

array([[999,  41,  28],
       [ 37,  46,  35],
       [ 29,  44,  51],
       [ 42,  39,  33],
       [ 48,  52,  36]])

In [30]:
print(bloque)
print(viajes)

[[999  41]
 [ 37  46]]
[[999  41  28]
 [ 37  46  35]
 [ 29  44  51]
 [ 42  39  33]
 [ 48  52  36]]


33. Indeed this type of "change in the subset determines a change in the set" behavior that confused me deeply before. It seems that there are several mechanisms to why this can happen, but in general, it is simply that numpy instead of creating a whole new copy of an array (which can be extremely expensive to the memory) it simply provides a view of the whole array. I do wonder how could I create a literal copy of an array though

In [31]:
viajes[0, 0] = previous_value
viajes

array([[32, 41, 28],
       [37, 46, 35],
       [29, 44, 51],
       [42, 39, 33],
       [48, 52, 36]])

In [33]:
seleccion = viajes[viajes > 45]
seleccion

array([46, 51, 48, 52])

In [34]:
seleccion[0] = 999
print(seleccion)
print(viajes)

[999  51  48  52]
[[32 41 28]
 [37 46 35]
 [29 44 51]
 [42 39 33]
 [48 52 36]]


34. First of all, `viajes[viajes > 45]` didn't had the behavior I thought it would. I expected a `(5, 3)` matrix (just like the shape of `viajes`) that had only 0s and 1s that represented how each value was compared to `45` and whether it was false or true that they were larger, in other words I expected

```python
[[0 0 0]
 [0 1 0]
 [0 0 1]
 [0 0 0]
 [1 1 0]]
```

Why did I get a 1D array with all the values greater than 45 instead of the previous matrix? 

35. To be honest I am not sure why the origin doesn't change, but I believe that it is mainly for the reason of the shape change. While a new variable of a subsection is literally a view, with `viajes[viajes > 45]` I am asking a deliberate modification of the data, so in this case it makes sense that I shouldn't modify the origin, as we're talking about an array that completely changed its form and meaning. 

In [35]:
subset = viajes[:2, :2]
snd_selection = subset[subset < 40]
snd_selection

array([32, 37])

In [37]:
snd_selection[:] = 0
print(snd_selection)
print(subset)
print(viajes)

[0 0]
[[32 41]
 [37 46]]
[[32 41 28]
 [37 46 35]
 [29 44 51]
 [42 39 33]
 [48 52 36]]


37. Indeed in this case while I modified the selection it didn't affect its "parent" (`subset`) nor its "grandparent" (`viajes`). 

## Exercise 5

In [38]:
first_row = viajes[3, 2]
print(first_row)
print(first_row.shape)

33
()


38. It selects the fourth row and third column. It has no shape as it is a scalar, it is the data of route C in Thursday and the view is simply 33

In [39]:
second_row = viajes[1:3, 1:]
print(second_row)
print(second_row.shape)

[[46 35]
 [44 51]]
(2, 2)


39. It selects the second and third columns and rows. Its shape is `(2, 2)`. It is the data from routes B and C from Tuesday to Wednesday and its view is 

```python
[[46 35]
 [44 51]]
```

40. `viajes[viajes > 45]` selects each of the values of the matrix that are greater than 45. Its shape is `(1, )`, it has no axes as it is a 1D vector and its view is `[46, 51, 48, 52]`. Now to test it out. 

In [41]:
viajes[viajes > 45]
viajes

array([[32, 41, 28],
       [37, 46, 35],
       [29, 44, 51],
       [42, 39, 33],
       [48, 52, 36]])

41. `viajes[[4, 0, 2]]` selects the 4, 0 and 1 rows on that specific order. Its shape is `(3, 3)` and the context is that they're the Friday, Monday and Wednesday rows with all the 3 routes. The view would be
```python
[[48, 52, 36], 
[32, 41, 28], 
[29, 44, 51]]
```

Now to test it out

In [42]:
viajes[[4, 0, 2]]

array([[48, 52, 36],
       [32, 41, 28],
       [29, 44, 51]])

42. Indeed it was. This is the final of the first session and I use `.copy()` when I want to have a complete copy of the array in order to modify without cascading the changes to the original array. 